File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/skin/FluentToggleButtonSkin.java
package com.mcreater.fxfluent.controls.skin;

import com.mcreater.fxfluent.brush.AbstractColorBrush;
import com.mcreater.fxfluent.brush.SolidColorBrush;
import com.mcreater.fxfluent.controls.FluentToggleButton;
import com.mcreater.fxfluent.controls.state.StateType;
import com.mcreater.fxfluent.controls.value.AnimatedValue;
import com.mcreater.fxfluent.util.BrushUtil;
import com.mcreater.fxfluent.util.listeners.NewValueListener;
import com.sun.javafx.scene.control.skin.ToggleButtonSkin;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ObservableValue;
import javafx.geometry.Insets;
import javafx.scene.layout.CornerRadii;
import javafx.scene.paint.Color;
import javafx.util.Duration;

import java.util.stream.Stream;

import static com.mcreater.fxfluent.controls.state.StateUtil.genState;

public class FluentToggleButtonSkin extends ToggleButtonSkin {
    private final FluentToggleButton button;
    private final ObjectProperty<StateType> state = new SimpleObjectProperty<>(null);
    private final AnimatedValue<AbstractColorBrush> backgroundColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(83));
    private final AnimatedValue<AbstractColorBrush> foregroundColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(42));

    private final AnimatedValue<AbstractColorBrush> upBorderColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(83));
    private final AnimatedValue<AbstractColorBrush> downBorderColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(83));
    private final AnimatedValue<AbstractColorBrush> leftBorderColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(83));
    private final AnimatedValue<AbstractColorBrush> rightBorderColor = new AnimatedValue<>(new SolidColorBrush(Color.TRANSPARENT), Duration.millis(83));
    public FluentToggleButtonSkin(FluentToggleButton control) {
        super(control);
        button = control;

        Stream.of(
                control.hoverProperty(),
                control.pressedProperty(),
                control.focusedProperty(),
                control.disabledProperty()
        ).forEach(a -> a.addListener(this::updateState));

        button.selectedProperty().addListener((NewValueListener<Boolean>) t1 -> FluentToggleButtonSkin.this.updateComponents(state.get()));

        CornerRadii cornerRadii = control.getCornerRadii();

        state.addListener((NewValueListener<StateType>) FluentToggleButtonSkin.this::updateComponents);
        backgroundColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.backgroundFill(cornerRadii))
        );
        foregroundColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.textFill())
        );
        upBorderColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.borderFill(BrushUtil.BorderOrientation.TOP, cornerRadii))
        );
        downBorderColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.borderFill(BrushUtil.BorderOrientation.BOTTOM, cornerRadii))
        );
        leftBorderColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.borderFill(BrushUtil.BorderOrientation.LEFT, cornerRadii))
        );
        rightBorderColor.getProperty().addListener((NewValueListener<AbstractColorBrush>) newValue ->
                newValue.accept(this.button, BrushUtil.borderFill(BrushUtil.BorderOrientation.RIGHT, cornerRadii))
        );

        updateState(null, null, null);
        updateComponents(state.get());
        button.setPadding(new Insets(5, 11, 5, 11));
    }

    private void updateState(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
        state.set(genState(button.isDisabled(), button.isHover(), button.isPressed(), button.isFocused()));
    }

    private void updateComponents(StateType type) {
        backgroundColor.updateValue(button.getBackgroundRemap().get(type).apply(button.getResourceDict()));
        foregroundColor.updateValue(button.getForegroundRemap().get(type).apply(button.getResourceDict()));
        Stream.of(
                upBorderColor, leftBorderColor, rightBorderColor
        ).forEach(a -> a.updateValue(button.getBorderRemap().get(StateType.PRESS).apply(button.getResourceDict())));
        downBorderColor.updateValue(button.getBorderRemap().get(type).apply(button.getResourceDict()));
    }

    public void implUpdate() {
        this.updateComponents(state.get());
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/skin/FluentToggleButtonSkin.kt
package com.mcreater.fxfluent.controls.skin

import com.mcreater.fxfluent.brush.AbstractColorBrush
import com.mcreater.fxfluent.brush.SolidColorBrush
import com.mcreater.fxfluent.controls.FluentToggleButton
import com.mcreater.fxfluent.controls.state.StateType
import com.mcreater.fxfluent.controls.state.StateUtil.Companion.genState
import com.mcreater.fxfluent.controls.value.AnimatedValue
import com.mcreater.fxfluent.util.BrushUtil
import com.mcreater.fxfluent.util.BrushUtil.Companion.backgroundFill
import com.mcreater.fxfluent.util.BrushUtil.Companion.borderFill
import com.mcreater.fxfluent.util.BrushUtil.Companion.textFill
import com.mcreater.fxfluent.util.listeners.NewValueListener
import com.sun.javafx.scene.control.skin.ToggleButtonSkin
import javafx.beans.property.ObjectProperty
import javafx.beans.property.ReadOnlyBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ObservableValue
import javafx.geometry.Insets
import javafx.scene.paint.Color
import javafx.util.Duration
import java.util.stream.Stream


class FluentToggleButtonSkin(private val button: FluentToggleButton) : ToggleButtonSkin(
    button
) {
    private val state: ObjectProperty<StateType?> = SimpleObjectProperty(null)
    private val backgroundColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(83.0))
    private val foregroundColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(42.0))
    private val upBorderColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(83.0))
    private val downBorderColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(83.0))
    private val leftBorderColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(83.0))
    private val rightBorderColor =
        AnimatedValue<AbstractColorBrush>(SolidColorBrush(Color.TRANSPARENT), Duration.millis(83.0))

    init {
        Stream.of(
            button.hoverProperty(),
            button.pressedProperty(),
            button.focusedProperty(),
            button.disabledProperty()
        ).forEach { a: ReadOnlyBooleanProperty ->
            a.addListener { _: ObservableValue<out Boolean?>?, _: Boolean?, _: Boolean? ->
                updateState()
            }
        }
        button.selectedProperty().addListener(NewValueListener {
            updateComponents(
                state.get()!!
            )
        })
        val cornerRadii = button.cornerRadii
        state.addListener(NewValueListener { updateComponents(it!!) })
        backgroundColor.property.addListener(
            NewValueListener { it.accept(button, backgroundFill(cornerRadii)) }
        )
        foregroundColor.property.addListener(
            NewValueListener { it.accept(button, textFill()) }
        )
        upBorderColor.property.addListener(
            NewValueListener { it.accept(button, borderFill(BrushUtil.BorderOrientation.TOP, cornerRadii)) }
        )
        downBorderColor.property.addListener(
            NewValueListener { it.accept(button, borderFill(BrushUtil.BorderOrientation.BOTTOM, cornerRadii)) }
        )
        leftBorderColor.property.addListener(
            NewValueListener { it.accept(button, borderFill(BrushUtil.BorderOrientation.LEFT, cornerRadii)) }
        )
        rightBorderColor.property.addListener(
            NewValueListener { it.accept(button, borderFill(BrushUtil.BorderOrientation.RIGHT, cornerRadii)) }
        )
        updateState()
        updateComponents(state.get()!!)
        button.padding = Insets(5.0, 11.0, 5.0, 11.0)
    }

    private fun updateState() {
        state.set(genState(button.isDisabled, button.isHover, button.isPressed, button.isFocused))
    }

    private fun updateComponents(type: StateType) {
        backgroundColor.updateValue(button.backgroundRemap[type]!!.apply(button.resourceDict)!!)
        foregroundColor.updateValue(button.foregroundRemap[type]!!.apply(button.resourceDict)!!)
        Stream.of(
            upBorderColor, leftBorderColor, rightBorderColor
        ).forEach { a: AnimatedValue<AbstractColorBrush> ->
            a.updateValue(
                button.borderRemap[StateType.PRESS]!!.apply(button.resourceDict)!!
            )
        }
        downBorderColor.updateValue(button.borderRemap[type]!!.apply(button.resourceDict)!!)
    }

    fun implUpdate() {
        updateComponents(state.get()!!)
    }
}
